### PR TITLE
feat(discover): added snql support for percentile functions

### PR DESCRIPTION
**Pull request by edwardgou-sentry**
Adds snql support for `percentile`, `p50`, `p75`, `p95`, `p99`, and `p100`

### CODE REVIEW

File: src/sentry/search/events/fields.py [hunk 1]
```diff
@@ -2759,6 +2809,39 @@ def _resolve_user_misery_function(self, args: Mapping[str, str], alias: str) ->
...
+                    Function(
+                        "divide",
+                        [
+                            self.column("measurements.frames_slow"),
+                            self.column("measurements.frames_total"),
+                        ],
+                    ),
+                    0,
+                ],
+            )
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:2838]: From a Slack conversation, this block of code doesn't really belong here but rather in an Argument resolver. We need to figure out a nicer way to implement it.

Comment (shruthilayaj) [reply src/sentry/search/events/fields.py:2838]: Going to move the measurement columns to `field_alias_converter`

Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: 👍 we've moved columns to `field_alias_converter`!

File: src/sentry/search/events/fields.py [hunk 2]
```diff
@@ -2763,6 +2826,43 @@ def _resolve_user_misery_function(self, args: Mapping[str, str], alias: str) ->
...
+                Function(
+                    "divide",
+                    [
+                        self.column(dividend),
+                        self.column(divisor),
+                    ],
+                ),
+                0,
+            ],
+        )
```
Comment (Zylphrex) [inline src/sentry/search/events/fields.py:2855]: A few things:
1. `_resolve_measurements` isn't a good name here because it's resolving 3 field aliases that happen to be prefixed with `measurement.`
2. The default value for the else case should be `None` instead of `0`.

File: src/sentry/search/events/fields.py [hunk 3]
```diff
@@ -1314,6 +1314,13 @@ def with_default(default, argument):
     return argument
 
 
+class SnQLColumnOrAlias(FunctionArg):
+    def normalize(self, value, _):
+        if value is None:
+            raise InvalidFunctionArgument("a column or alias is required")
+        return value
```
Comment (Zylphrex) [inline src/sentry/search/events/fields.py:1321]: This should more closely resemble the existing `NumericColumn` and `NumericColumnNoLookup`.

I believe that we've agreed on that the `normalize` method here should just do the necessary validations without resolving the column into `SnQL` and letting that resolution happen within `resolve_function` like you have below.

File: src/sentry/search/events/fields.py [hunk 4]
```diff
@@ -2265,13 +2275,63 @@ def __init__(self, dataset: Dataset, params: ParamsType):
...
+                    redundant_grouping=True,
+                ),
+                SnQLFunction(
+                    "p100",
+                    optional_args=[
+                        with_default("transaction.duration", SnQLColumnOrAlias("column")),
+                    ],
+                    snql_aggregate=lambda args, alias: self._resolve_percentile(args, alias, 1),
+                    default_result_type="integer",
+                    redundant_grouping=True,
```
Comment (Zylphrex) [inline src/sentry/search/events/fields.py:2332]: All these functions should support `reflexive_result_type()` eventually, not just `percentile`.

File: src/sentry/search/events/fields.py [hunk 5]
```diff
@@ -2763,6 +2826,43 @@ def _resolve_user_misery_function(self, args: Mapping[str, str], alias: str) ->
             alias,
         )
 
+    def _resolve_percentile(
+        self, args: Mapping[str, str], alias: str, fixed_percentile: float = None
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:2830]: I think the `Mapping` type might be incorrect- since the value can be either string, Column, SelectType, integer, float etc... ugh but I see that it's `Mapping[str, str]` everywhere else too. Maybe this can be cleaned up in a follow up.

File: src/sentry/search/events/fields.py (hunk 3 above)
Comment (shruthilayaj) [reply src/sentry/search/events/fields.py:1321]: And just to keep it consistent, you can just use the same name as the classes you are trying to replicate and prefix it with SnQL. So `SnQLNumericColumnNoLookup`, that will help clean up after the migration

File: src/sentry/search/events/fields.py [hunk 6]
```diff
@@ -2265,13 +2275,63 @@ def __init__(self, dataset: Dataset, params: ParamsType):
...
                     default_result_type="percentage",
                 ),
+                SnQLFunction(
+                    "percentile",
+                    required_args=[
+                        SnQLColumnOrAlias("column"),
+                        NumberRange("percentile", 0, 1),
+                    ],
+                    snql_aggregate=self._resolve_percentile,
+                    result_type_fn=reflective_result_type(),
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:2285]: You need a `get_type` [method](https://github.com/getsentry/sentry/blob/master/src/sentry/search/events/fields.py#L1219) on your subclass and support [this](https://github.com/getsentry/sentry/blob/master/src/sentry/search/events/fields.py#L1535) I think. So if you do p95 on an integer type field, it will return an integer type and if you do it on a a duration type field (like `transaction.duration` it will result in a duration type.

File: src/sentry/search/events/fields.py [hunk 7]
```diff
@@ -2265,13 +2275,63 @@ def __init__(self, dataset: Dataset, params: ParamsType):
...
+                    default_result_type="duration",
+                    redundant_grouping=True,
+                ),
+                SnQLFunction(
+                    "p50",
+                    optional_args=[
+                        with_default("transaction.duration", SnQLColumnOrAlias("column")),
+                    ],
+                    snql_aggregate=lambda args, alias: self._resolve_percentile(args, alias, 0.5),
+                    default_result_type="integer",
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:2295]: if your default column is going to be `transaction.duration` the default result type should probably be `duration`

Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: Fixed!

File: src/sentry/search/events/fields.py (hunk 6 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py:2306]: I've added `get_type` to  `SnQLNumericColumnNoLookup`! It's following the original `get_type` from `NumericColumnNoLookup`

File: src/sentry/search/events/fields.py (hunk 4 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: Updated to `reflective_result_type()`!

File: src/sentry/search/events/fields.py (hunk 3 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: I've renamed to `SnQLNumericColumnNoLookup` and updated to extend `NumericColumn`. `NumericColumnNoLookup` itself didn't have any validation, so just extending `NumericColumn` to inherit validation should be sufficient.

File: src/sentry/search/events/fields.py (hunk 2 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: 1. Renamed to `_resolve_division`, hopefully this is a better description since all it's doing is division.
2. Updated default to `None`!

File: src/sentry/search/events/fields.py [hunk 8]
```diff
@@ -1314,6 +1314,34 @@ def with_default(default, argument):
...
+
+    def normalize(self, value, params):
+        super().normalize(value, params)
+        return value
+
+    def get_type(self, value):
+        # `measurements.frames_frozen_rate` and `measurements.frames_slow_rate` are aliases
+        # to a percentage value, since they are expressions rather than columns, we special
+        # case them here
+        if isinstance(value, list):
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:1336]: This is JSON specific syntax, I don't think you'd ever get value type as list here

File: src/sentry/search/events/fields.py [hunk 9]
```diff
@@ -1314,6 +1314,34 @@ def with_default(default, argument):
...
+            for name in self.measurement_aliases:
+                field = FIELD_ALIASES[name]
+                expression = field.get_expression(None)
+                if expression == value:
+                    return field.result_type
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:1341]: We currently don't support result_type on the field aliases in the QueryFields. This reference to `FIELD_ALIASES` is the legacy JSON version that has the `PseudoFields`

Comment (shruthilayaj) [reply src/sentry/search/events/fields.py:1341]: Maybe we should just remove L1336-L1341 for now?

File: src/sentry/search/events/fields.py [hunk 10]
```diff
@@ -1314,6 +1314,34 @@ def with_default(default, argument):
...
+class SnQLNumericColumnNoLookup(NumericColumn):
+    measurement_aliases = {
+        MEASUREMENTS_FRAMES_SLOW_RATE,
+        MEASUREMENTS_FRAMES_FROZEN_RATE,
+        MEASUREMENTS_STALL_PERCENTAGE,
+    }
+
+    def __init__(self, name, allow_array_value=False):
+        super().__init__(name)
+        self.allow_array_value = allow_array_value
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:1326]: self.allow_array_value isn't being used, can probably be removed?

File: src/sentry/search/events/fields.py [hunk 11]
```diff
@@ -2483,8 +2569,8 @@ def resolve_function(self, function: str, match: Optional[Match[str]] = None) ->
 
         arguments = snql_function.format_as_arguments(name, arguments, self.params)
         for arg in snql_function.args:
-            if isinstance(arg, SnQLColumn):
-                arguments[arg.name] = self.resolve_field(arguments[arg.name])
+            if type(arg) in [SnQLNumericColumnNoLookup, SnQLColumn]:
+                arguments[arg.name] = self.resolve_column(arguments[arg.name])
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:2573]: is this change necessary? can it remain `self.resolve_field`?

File: src/sentry/search/events/fields.py (hunk 8 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: 👍  removed this condition!

File: src/sentry/search/events/fields.py (hunk 9 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: removed!

File: src/sentry/search/events/fields.py (hunk 10 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: Good point, removed!

File: src/sentry/search/events/fields.py (hunk 11 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py:2563]: Hey Shruthi, looks like there was a recent change to `resolve_field`. `resolve_function` and `resolve_field_alias` were moved out of `resolve_field` and into the new function `resolve_column` which also calls `resolve_field`. Since we would need to support the measurement aliases for percentile, I think we're required to use `resolve_column` for at least `SnQLNumericColumnNoLookup`. Maybe @wmak can comment on this, is this correct?

File: src/sentry/search/events/fields.py [hunk 12]
```diff
@@ -1314,6 +1314,24 @@ def with_default(default, argument):
     return argument
 
 
+class SnQLNumericColumnNoLookup(NumericColumn):
```
Comment (wmak) [inline src/sentry/search/events/fields.py:1317]: Could we just use Numeric Column here instead of introducing a new Column type? if this is for the type check whether we resolve args, I think we'll want to resolve anything that's a column not just the ones we've called out as `SnQL...Column`s

File: src/sentry/search/events/fields.py [hunk 13]
```diff
@@ -2841,6 +2917,46 @@ def _resolve_user_misery_function(self, args: Mapping[str, str], alias: str) ->
             alias,
         )
 
+    def _resolve_percentile(
+        self,
+        args: Mapping[str, Union[str, Column, SelectType, int, float]],
+        alias: str,
+        fixed_percentile: float = None,
+    ) -> SelectType:
+        return Function(
```
Comment (wmak) [inline src/sentry/search/events/fields.py:2926]: This will return slightly different results for p100, which is actually implemented with `max` instead. I believe this is because the quantile function is non-deterministic (especially on larger datasets) 
https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/reference/quantile/

File: src/sentry/search/events/fields.py (hunk 11 above)
Comment (wmak) [reply src/sentry/search/events/fields.py:2563]: can confirm, I've changed resolve_field to mean specifically that we're resolving a field (ie. we know what type the alias is already) compared to resolve_column where we're not usre if its an alias, column or field

File: src/sentry/search/events/fields.py (hunk 12 above)
Comment (shruthilayaj) [reply src/sentry/search/events/fields.py:1317]: I don't think we can just use NumericColumn here since because of this return [statement](https://github.com/getsentry/sentry/blob/master/src/sentry/search/events/fields.py#L1176). We'd just want to return the original value (which is the alias) here and not the mapped Discover column. The `self.column()` call in the QueryFields class will handle that

File: src/sentry/search/events/fields.py [hunk 14]
```diff
@@ -1314,6 +1314,24 @@ def with_default(default, argument):
     return argument
 
 
+class SnQLNumericColumnNoLookup(NumericColumn):
+    measurement_aliases = {
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:1318]: Doesn't seem like this is being used anymore

File: src/sentry/search/events/fields.py [hunk 15]
```diff
@@ -1314,6 +1314,24 @@ def with_default(default, argument):
...
+    }
+
+    def __init__(self, name):
+        super().__init__(name)
+
+    def normalize(self, value, params):
+        super().normalize(value, params)
+        return value
+
+    def get_type(self, value):
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:1331]: If all you are doing is calling super on get_type, you don't need this method. It will fallback to the method on the superclass

File: src/sentry/search/events/fields.py [hunk 16]
```diff
@@ -1314,6 +1314,24 @@ def with_default(default, argument):
...
 
 
+class SnQLNumericColumnNoLookup(NumericColumn):
+    measurement_aliases = {
+        MEASUREMENTS_FRAMES_SLOW_RATE,
+        MEASUREMENTS_FRAMES_FROZEN_RATE,
+        MEASUREMENTS_STALL_PERCENTAGE,
+    }
+
+    def __init__(self, name):
```
Comment (shruthilayaj) [inline src/sentry/search/events/fields.py:1324]: Since the init isn't doing anything specific to this class (ie, it's just calling `super`) you can remove this method as well

**Review by shruthilayaj — COMMENTED** [review]
Minor comments

File: src/sentry/search/events/fields.py (hunk 13 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: 👍  added a condition to use `max` when querying for p100

File: src/sentry/search/events/fields.py (hunk 14 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: removed!

File: src/sentry/search/events/fields.py (hunk 15 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: removed!

File: src/sentry/search/events/fields.py (hunk 16 above)
Comment (edwardgou-sentry) [reply src/sentry/search/events/fields.py]: removed!

File: src/sentry/search/events/fields.py (hunk 12 above)
Comment (wmak) [reply src/sentry/search/events/fields.py:1317]: Hmm I still think we don't need to introduce new Columns here but let me try that in a separate PR so we can get this change in :+1:

**Review by wmak — APPROVED** [review]

**Review by shruthilayaj — APPROVED** [review]

**Comment by edwardgou-sentry** [comment]
@shruthilayaj @Zylphrex @wmak Thanks for the reviews! 🥇
